test/cpp/tensorexpr/test_loopnest.cpp

```diff
@@ -546,5 +547,224 @@ void testScheduleDynamicShape2D() {
+  // For this loop bounds inference should yield the following:
+  // {{b, kStore, 0, 99}, {a, kLoad, 0, 99}}
+  KernelScope kernel_scope;
+  ExprHandle n(100);
+  Buffer a(BufHandle("a", {n}), kFloat);
+  Tensor* b =
+      Compute("b", {{n, "i"}}, [&](const VarHandle& i) { return a(i); });
+  LoopNest l({b});
+  const std::vector<TensorAccessBoundsInfo>& bounds_info =
+      inferBounds(l.root_stmt());
+  auto bounds_info_map = convertBoundsInfoToMap(bounds_info);
+
+  // We should have two entries: one for 'b' and one for 'a'.
+  ASSERT_EQ(bounds_info_map.size(), 2);
+  ASSERT_EQ(bounds_info_map.at(a.data()).kind, kLoad);
+  verifyConstBounds(bounds_info_map.at(a.data()), {{0, 99}});
+  ASSERT_EQ(bounds_info_map.at(b->buf()).kind, kStore);
+  verifyConstBounds(bounds_info_map.at(b->buf()), {{0, 99}});
+}
+
+void testBoundsInference_2() {
+  // Verify that bounds inference works for the following example:
+  // for i in 0..n:
+  //   b[i] = a[i]
+  // For this loop bounds inference should yield the following:
+  // {{b, kStore, 0, n-1}, {a, kLoad, 0, n-1}}
+  KernelScope kernel_scope;
+  VarHandle n("n", kInt);
+  Buffer a(BufHandle("a", {n}), kFloat);
+  Tensor* b =
+      Compute("b", {{n, "i"}}, [&](const VarHandle& i) { return a(i); });
+  LoopNest l({b});
+  const std::vector<TensorAccessBoundsInfo>& bounds_info =
+      inferBounds(l.root_stmt());
+  auto bounds_info_map = convertBoundsInfoToMap(bounds_info);
+
+  // We should have two entries: one for 'b' and one for 'a'.
+  ASSERT_EQ(bounds_info_map.size(), 2);
+  ASSERT_EQ(bounds_info_map.at(a.data()).kind, kLoad);
+  verifyConstBounds(bounds_info_map.at(a.data()), {{0, -1}});
+  ASSERT_EQ(bounds_info_map.at(b->buf()).kind, kStore);
+  verifyConstBounds(bounds_info_map.at(b->buf()), {{0, -1}});
+}
+
+void testBoundsInference_3() {
+  // Verify that bounds inference works for the following example:
+  // for i in 0..100:
+  //   b[i] = a[i] * a[i+10]
+  // For this loop bounds inference should yield the following:
+  // {{b, kStore, 0, 99}, {a, kLoad, 0, 109}}
+  KernelScope kernel_scope;
+  ExprHandle n(100);
+  Buffer a(BufHandle("a", {n + 10}), kFloat);
+  Tensor* b = Compute(
+      "b", {{n, "i"}}, [&](const VarHandle& i) { return a(i) * a(i + 10); });
+  LoopNest l({b});
+  const std::vector<TensorAccessBoundsInfo>& bounds_info =
+      inferBounds(l.root_stmt());
+  auto bounds_info_map = convertBoundsInfoToMap(bounds_info);
+
+  // We should have two entries: one for 'b' and one for 'a'.
+  ASSERT_EQ(bounds_info_map.size(), 2);
+  ASSERT_EQ(bounds_info_map.at(a.data()).kind, kLoad);
+  verifyConstBounds(bounds_info_map.at(a.data()), {{0, 109}});
+  ASSERT_EQ(bounds_info_map.at(b->buf()).kind, kStore);
+  verifyConstBounds(bounds_info_map.at(b->buf()), {{0, 99}});
+}
+
+void testBoundsInference_4() {
+  // Verify that bounds inference works for the following example:
+  //
+  // for y in 0..200:
+  //   for x in 0..320:
+  //     b[y,x] = x*y
+  // for y in 0..200:
```

Comment:
An important case is where the "y, x" in "c" uses a small range in "a" and "b".

Reply:
Added!